### PR TITLE
WIP: Add dbscan clustering method

### PR DESCRIPTION
**Pull request by abieler**
First shot at implementing DBSCAN in nim.
Algorithm from sklearn

- [x] `debug.nim` to be removed
- [x]  `moons.csv` to be removed
- [x]  fix bug for the last record of the data tensor
- [ ] replace brute force nearest neighbor calculation with something smarter
- [ ] tests to be added

Basic usage:
```nim
import dbscan

let X = toTensor(...)
let metric = Euclidean()
let labels = dbscan(X, 0.25, 5, metric)
```

Please weigh in on my preliminary design decision with the `metrics` and the use of multimethods for the `pairwise()` function.

FYI: nim newbie here, so mostly having fun / toying around with the language so far.

Cheers

### CODE REVIEW

**Comment by mratsim** [comment]
Thanks, this looks quite simple and maintainable.

#### Regarding methods:
- you're actually using single dispatch since the only polymorphic part is the `Metric`
- generic methods are in bad shape in Nim and Tensor[float] is generic over float32 and float64:
  - https://forum.nim-lang.org/t/4921
  - https://github.com/nim-lang/Nim/issues/6283
  - I had to remove them from Arraymancer
    - https://github.com/mratsim/Arraymancer/issues/327 -> https://github.com/mratsim/Arraymancer/pull/333
- minor: since method dispatch at runtime, the `pairwise` will always end-up in the final binary even if unused (no dead code elimination).

#### Alternative 1 - enum-dispatch (not advised)

Use an static enum for dispatching, for example replace
```Nim
type
  Metric* = ref object of RootObj
  Euclidean* = ref object of Metric
  Manhattan* = ref object of Metric
  Minkowski* = ref object of Metric
    p: int
  Jaccard* = ref object of Metric


method pairwise*(x, y: Tensor[float], metric: Metric): float {.base.} =
  return -1

method pairwise*(x, y: Tensor[float], metric: Euclidean): float =
  var d: float
  for v, w in zip(x, y):
    d = (v - w)
    result +=  d * d
  result = sqrt(result)


method pairwise*(x, y: Tensor[float], metric: Manhattan): float =
  for v, w in zip(x, y):
        result += abs(v - w)


method pairwise*(x, y: Tensor[float], p:float, metric: Minkowski): float =
    var d: float
    for v, w in zip(x, y):
        d = abs(float(v - w)) 
        result += pow(d, p)
    result = pow(result, 1/p)
```

by

```Nim
type
  Metric* = enum
    Euclidean
    Manhattan
    Minkowski
    Jaccard

proc pairwiseEuclidean(x, y: Tensor[float]): float =
  var d: float
  for v, w in zip(x, y):
    d = (v - w)
    result +=  d * d
  result = sqrt(result)


proc pairwiseManhattan(x, y: Tensor[float]): float =
  for v, w in zip(x, y):
    result += abs(v - w)


proc pairwiseMinkowski(x, y: Tensor[float], p:float): float =
  var d: float
  for v, w in zip(x, y):
      d = abs(float(v - w)) 
      result += pow(d, p)
  result = pow(result, 1/p)

proc pairwise[T](x, y: Tensor[T], metric: static Metric, p = T(NaN)): float =
  when metric == Euclidean:
    pairwiseEuclidean(x, y)
  elif metric == Manhattan:
    pairwiseManhattan(x, y)
  elif metric == Minkowski:
    pairwiseMinkowski(x, y, p)
  else:
    {.error: "Unreachable".}
```

This works because I expect the enum to not grow wildly in the future hence we don't suffer much from the Expression Problem which in short is:
- adding new types to enum-based dispatch is hard because you need to update all **existing** functions while adding new functions is easy, you should dispatch on all existing types and don't need to update existing code.
- adding new functions/interface to method-based dispatch is hard because you need to create one for each **existing** types (even with NotImplementedError) while adding new type is easy, create the type and all relevant methods

i.e. if you need to add a new function/types, can you do it in a self-contained extra file or do you need to edit existing code or reformulated: is it easy to add operations or datatypes.

#### Alternative 2 - typedesc dispatch (advised)

Nim offers almost dependent-types and they can be used for compile-time dispatch. They are IMO much nicer to work with than static enums in many-cases.

```Nim
type
  Euclidean* = object
  Manhattan* = object
  Minkowski* = object
  Jaccard* = object

proc pairwise(metric: typedesc[Euclidean], x, y: Tensor[float]): float =
  var d: float
  for v, w in zip(x, y):
    d = (v - w)
    result +=  d * d
  result = sqrt(result)

proc pairwise(metric: typedesc[Manhattan], x, y: Tensor[float]): float =
  for v, w in zip(x, y):
    result += abs(v - w)

proc pairwise(metric: typedesc[Minkowski], x, y: Tensor[float], p:float): float =
  var d: float
  for v, w in zip(x, y):
      d = abs(float(v - w)) 
      result += pow(d, p)
  result = pow(result, 1/p)
```

You get the best of both world:
- Adding a new type, doen't require to change a when/if/case statement in all existing function
- Adding a new function to a datatype doesn't require you to add one for each types

As a side-effect, this doesn't use the name `Metric`, there might be more obvious use of this name so keeping this option open for later is good.

**Comment by abieler** [comment]
thanks for your valuable input! as i am trying out the `typedesc` version i noticed that
before i would be able to do:

```nim
let metric = Euclidean()
let X = toTensor(....)
var result = dbscan(X, ...., metric)
# or
let metric = Minkowski(p=4)
var result = dbscan(X, ...., metric)
```

so the api for `dbscan()` remains the same, independent of the number of parameters of the metric.
Now with `typedesc` I think I would have to do

```nim
let X = toTensor(....)
var result = dbscan(X, ...., Minkowski, p=4)
# or
var result = dbscan(X, ...., Manhattan)
```
hence come up with two implementations of `dbscan()`. am I missing something here?


cheers
andre

File: src/ml/cluster/dbscan.nim [hunk 1]
```diff
@@ -0,0 +1,78 @@
+import algorithm
+import deques
+import sequtils
+
+import ../metrics/distance
+import ../../tensor/tensor
+
+
+proc argsort[T: SomeFloat](a: seq[T]): seq[int] =
+  result = toSeq(0..<high(a))
```
Comment (brentp) [inline src/ml/cluster/dbscan.nim]: should this be `toSeq(0..high(a))` ?

Comment (mratsim) [reply src/ml/cluster/dbscan.nim]: yes or `toSeq(0..<a.len)`

**Comment by mratsim** [comment]
You can have:

```Nim
type
  Euclidean* = object
  Manhattan* = object
  Minkowski* = object
  Jaccard* = object

proc dbscan*[T: SomeFloat](X: Tensor[T], eps: float, minSamples: int, metric: typedesc[Euclidean or Manhattan or Jaccard]): seq[int] =
    let neighborhoods = computeNeighborhoods(X, eps, metric)
    let is_core = findCorePoints(neighborhoods, minSamples)
    result = assignLabels(is_core, neighborhoods)

proc dbscan*[T: SomeFloat](X: Tensor[T], eps: float, minSamples: int, metric: typedesc[Minkowski], p: int): seq[int] =
    let neighborhoods = computeNeighborhoods(X, eps, metric, p)
    let is_core = findCorePoints(neighborhoods, minSamples)
    result = assignLabels(is_core, neighborhoods)
```

which seems reasonable in terms of duplication.

**Comment by Vindaar** [comment]
@abieler any chance you'll continue this PR in the near future? I might want to give this a try for my work soonish (meaning in a few weeks probably). If not I'll probably pick up from your PR at that point, if that's ok with you.

**Comment by abieler** [comment]
Hey @Vindaar I ll give it another crack over the Easter break

**Comment by abieler** [comment]
@Vindaar  or anyone feel free to hijack this PR. i unfortunately won't have a lot of time to invest here :(

the code is working and i think ok efficient except for the neighborhood computation which is a naive linear search. replacing it with kdtree or so would definitely be a huge improvement.

**Comment by mratsim** [comment]
Thanks @abieler

**Comment by Vindaar** [comment]
No worries and thanks for your work!

I started a kd tree implementation, which is like half done in arraymancer recently for something else. I might rebase my code on your PR and use try to make use of that.

**Comment by Vindaar** [comment]
With #518 merged, I'll close this.

Thanks again @abieler!

**Pull request closed without merging**